velox/type/DecimalUtil.cpp

```diff
@@ -49,11 +49,113 @@ std::string formatDecimal(uint8_t scale, int128_t unscaledValue) {
   return fmt::format(
       "{}{}{}", isNegative ? "-" : "", integralPart, fractionString);
 }
+
+int32_t getBitCount(uint32_t i) {
+  static constexpr int kMaxBits = std::numeric_limits<uint64_t>::digits;
+  uint64_t num = static_cast<uint32_t>(i);
+  return bits::countBits(reinterpret_cast<uint64_t*>(&num), 0, kMaxBits);
+}
+
+// Origins from java side BigInteger#bitLength.
+//
+// Returns the number of bits in the minimal two's-complement
+// representation of this BigInteger, <em>excluding</em> a sign bit.
+// For positive BigIntegers, this is equivalent to the number of bits in
+// the ordinary binary representation.  For zero this method returns
+// {@code 0}.  (Computes {@code (ceil(log2(this < 0 ? -this : this+1)))}.)
+//
+// @return number of bits in the minimal two's-complement
+//         representation of this BigInteger, <em>excluding</em> a sign bit.
+int32_t getBitLength(int8_t sig, const std::vector<int32_t>& mag) {
+  int32_t len = mag.size();
+  int32_t n = -1;
+  if (len == 0) {
+    n = 0;
+  } else {
+    // Calculate the bit length of the magnitude.
+    int32_t magBitLength = ((len - 1) << 5) + 64 -
+        bits::countLeadingZeros(static_cast<uint64_t>(mag[0]));
+    if (sig < 0) {
+      // Check if magnitude is a power of two.
+      bool pow2 = (getBitCount((uint32_t)mag[0]) == 1);
+      for (int i = 1; i < len && pow2; ++i) {
+        pow2 = (mag[i] == 0);
+      }
+      n = (pow2 ? magBitLength - 1 : magBitLength);
+    } else {
+      n = magBitLength;
+    }
+  }
+  return n;
+}
+
+// The magnitude of this BigInteger, in <i>big-endian</i> order: the
+// zeroth element of this array is the most-significant int of the
+// magnitude.  The magnitude must be "minimal" in that the most-significant
+// int ({@code mag[0]}) must be non-zero.  This is necessary to
+// ensure that there is exactly one representation for each BigInteger
+// value.  Note that this implies that the BigInteger zero has a
+// zero-length mag array.
+std::vector<int32_t> convertToIntMags(uint128_t value) {
+  std::vector<int32_t> mag;
+  int32_t v1 = value >> 96;
+  int32_t v2 = value >> 64;
+  int32_t v3 = value >> 32;
+  int32_t v4 = value;
+  if (v1 != 0) {
+    mag.emplace_back(v1);
+    mag.emplace_back(v2);
+    mag.emplace_back(v3);
+    mag.emplace_back(v4);
+  } else if (v2 != 0) {
+    mag.emplace_back(v2);
+    mag.emplace_back(v3);
+    mag.emplace_back(v4);
+  } else if (v3 != 0) {
+    mag.emplace_back(v3);
+    mag.emplace_back(v4);
+  } else if (v4 != 0) {
+    mag.emplace_back(v4);
+  }
+  return mag;
+}
 } // namespace
 
 std::string DecimalUtil::toString(const int128_t value, const TypePtr& type) {
   auto [precision, scale] = getDecimalPrecisionScale(*type);
   return formatDecimal(scale, value);
 }
 
+int32_t DecimalUtil::getByteArrayLength(int128_t value) {
```

Comment:
No, because we use `(len - 1) >>5` to compute the magBitLength, so it should be int32_t mag vector.
https://github.com/facebookincubator/velox/pull/5444/files#diff-b395c0d4e1d7961eb3b34688ebf64eafe59b0d4c0f9a8809cf04eb82f7c36dfdR76

Reply:
Yes, simplified.